### PR TITLE
[WW-5370] Makes HttpParameters case-insensitive

### PR DESCRIPTION
**Pull request by lukaszlenart**
Closes [WW-5370](https://issues.apache.org/jira/browse/WW-5370)

### CODE REVIEW

**Review by yasserzamani — APPROVED** [review]
LGTM, thanks!

P.S. I'm not sure how much it's hard to implement but you might know better: Looks like that same functionality can be achieved with less code by replacing

final private Map<String, Parameter> parameters;

with

final private Map<String, Parameter> parameters = new TreeMap<String, Parameter>(String.CASE_INSENSITIVE_ORDER);

Or using Apache Commons [CaseInsensitiveMap](https://commons.apache.org/proper/commons-collections/apidocs/org/apache/commons/collections4/map/CaseInsensitiveMap.html).

**Comment by kusalk** [comment]
Yeah good call - would be better to change the underlying map to be one of the off-the-shelf implementations. The Apache Commons one probably makes more sense.

**Comment by lukaszlenart** [comment]
Thanks for the ideas! I decided to use `TreeMap` as proposed by Yasser to avoid introducing another dependency (Commons Collection is used only in tests)

File: core/src/main/java/org/apache/struts2/dispatcher/HttpParameters.java [hunk 1]
```diff
@@ -29,25 +29,27 @@
...
 
     public static Builder create() {
-        return new Builder(new HashMap<>());
+        return new Builder(new TreeMap<>(String.CASE_INSENSITIVE_ORDER));
     }
 
     public HttpParameters remove(Set<String> paramsToRemove) {
         for (String paramName : paramsToRemove) {
-            parameters.remove(paramName);
+            parameters.entrySet().removeIf(p -> p.getKey().equalsIgnoreCase(paramName));
```
Comment (kusalk) [inline core/src/main/java/org/apache/struts2/dispatcher/HttpParameters.java:52]: Do we still need this custom comparison?

File: core/src/main/java/org/apache/struts2/dispatcher/HttpParameters.java [hunk 2]
```diff
@@ -73,7 +78,14 @@ private Map<String, String[]> toMap() {
...
     }
 
+    /**
+     * Appends all the parameters by overriding any existing params in a case-insensitive manner
+     *
+     * @param newParams A new params to append
+     * @return a current instance of {@link HttpParameters}
+     */
     public HttpParameters appendAll(Map<String, Parameter> newParams) {
+        remove(newParams.keySet());
```
Comment (kusalk) [inline core/src/main/java/org/apache/struts2/dispatcher/HttpParameters.java:88]: Probably don't need this anymore either right?

File: core/src/main/java/org/apache/struts2/dispatcher/HttpParameters.java [hunk 3]
```diff
@@ -100,7 +112,7 @@ public boolean containsValue(Object value) {
 
     @Override
     public Parameter get(Object key) {
-        if (parameters.containsKey(key)) {
+        if (key != null && contains(String.valueOf(key))) {
```
Comment (kusalk) [inline core/src/main/java/org/apache/struts2/dispatcher/HttpParameters.java:115]: Not sure this part makes sense, checking for the existence of String#valueOf then retrieving using the object?

Do you mean something like this:
```
@Override
public Parameter get(Object key) {
    if (key == null) {
        return new Parameter.Empty("null");
    }
    Parameter val = parameters.get(key.toString());
    return val != null ? val : new Parameter.Empty(key.toString());
}
```

File: core/src/main/java/org/apache/struts2/dispatcher/HttpParameters.java (hunk 1 above)
Comment (lukaszlenart) [reply core/src/main/java/org/apache/struts2/dispatcher/HttpParameters.java]: Right, removed

File: core/src/main/java/org/apache/struts2/dispatcher/HttpParameters.java (hunk 3 above)
Comment (lukaszlenart) [reply core/src/main/java/org/apache/struts2/dispatcher/HttpParameters.java]: Simplified to return `null` - no idea what was the reasoning behind :D

Comment (lukaszlenart) [reply core/src/main/java/org/apache/struts2/dispatcher/HttpParameters.java]: I will use your code and I will add missing equals&hashCode

**Review by kusalk — APPROVED** [review]

**Review by yasserzamani — APPROVED** [review]
Thank you Lukasz! 🙇

**Pull request merged**